File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    /*
        Technique: Two Pointers

        1. Reverse the entire array.
        2. Reverese the first k elements.
        3. Reverse the remaining elements.

        TC: O(N) SC: O(1)
    */

    public void rotate(int[] nums, int k) {
        int N = nums.length;
        k = k % N;

        reverse(nums, 0, N - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, N - 1);
    }

    void reverse(int[] arr, int s, int e) {
        while (s < e) {
            int temp = arr[s];
            arr[s] = arr[e];
            arr[e] = temp;

            s++;
            e--;
        }
    }
}

File: 0189-rotate-array/0189-rotate-array.py
"""
Technique: Two Pinters
TC: O(N) SC: O(1)
"""


class Solution:
    def rotate(self, nums: list[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        N = len(nums)
        k = k % N
        self.reverse(nums, 0, N - 1)
        Solution.reverse(nums, 0, k - 1)
        Solution.reverse(nums, k, N - 1)

    @classmethod
    def reverse(cls, arr, s, e):
        while s < e:
            temp = arr[s]
            arr[s] = arr[e]
            arr[e] = temp
            s += 1
            e -= 1
